DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+apply to the argument.
+
+#### Copy constructor call vs. Blitting
+
+When a copy constructor is not defined for a struct, initializations are treated
+by copying the contents from the memory location of the right-hand side expression
+to the memory location of the left-hand side one (called "blitting"):
+
+```d
+struct A
+{
+    int[] a;
+}
+
+void main()
+{
+    A a = A([7]);
+    A b = a;                 // mempcy(&b, &a)
+    immutable A c = A([12]);
+    immutable A d = c;       // memcpy(&d, &c)
+}
+```
+
+Whenever a copy constructor is defined for a struct all implicit blitting is disabled for
+that struct:
+
+```d
+struct A
+{
+    int[] a;
+    this(ref A rhs) {}
+}
+
+void fun(immutable A) {}
+
+void main()
+{
+    immutable A a;
+    fun(a);          // error: copy constructor cannot be called with types (immutable) immutable
+}
+```
+
+#### Interaction with `alias this`
+
+There are situations in which a struct defines both an `alias this` and a copy constructor, and
+for which assignments to variables of the struct type may lead to ambiguities:
+
+```d
+struct A
+{
+    int a;
+    immutable(A) fun()
+    {
+        return immutable A(7);
+    }
+
+    alias fun this;
+
+    this(ref A another) immutable {}
+}
+
+struct B
+{
+    int a;
+    A fun()
+    {
+        return A(7);
+    }
+
+    alias fun this;
+
+    this(ref B another) immutable {}
+}
+
+void main()
+{
+    immutable A ia;
+    A a = ia;            // 1 - calls copy constructor
+
+    A bc;
+    A b = bc;            // 2 - calls B.fun
+}
+```
+
+In situations where both the copy constructor and `alias this` are suitable
+to solve the assignment (1), the copy constructor takes precedence over `alias this`
+because it is considered more specialized (the sole purpose of the copy constructor is to
+create copies). However, if no copy constructor in the overload set matches the exact
+qualified types of the source and the destination, the `alias this` is preferred (2).
+
+#### Generating copy constructors
+
+A copy constructor is generated for a `struct` S  if one of S's fields defines
+a copy constructor that S does not define.
+
+A field of a `struct` may define multiple copy constructors. In this situation,
+a copy constructor is generated for each overload.
+
+```d
+struct M
+{
+    this(ref immutable M) {}
+    this(ref immutable M) shared {}
+}
+
+struct S
+{
+    M m;
+    this(ref shared S) {}
+    /* The copy constructors with the following signatures are generated:
+        this(ref immutable S)
+        this(ref immutable S) shared
+     */
+}
+```
+
+The body of all the generated copy constructors does memberwise initialization for all
+the fields:
+
+```d
+{
+    this.field1 = s.field1;
+    this.field2 = s.field2;
+    ...;
+}
+```
+
+Inside the body of the generated copy constructors, for the fields that define a copy
+constructor, the assignment will be rewritten to a call to it; for those that do not,
```

Comment:
for the fields that define a copy constructor, the assignment will be rewritten to a call to it; -> the assignment will be rewritten as a call to the copy constructor of each field that defines one;